tpm2/error.go

```diff
@@ -0,0 +1,325 @@
+	RcReferenceS5:    "the 6th authorization session handle references a session that is not loaded",
+	RcReferenceS6:    "the 7th authorization session handle references a session that is not loaded",
+	RcNVRate:         "the TPM is rate-limiting accesses to prevent wearout of NV",
+	RcLockout:        "authorizations for objects subject to DA protection are not allowed at this time because the TPM is in DA lockout mode",
+	RcRetry:          "the TPM was not able to start the command",
+	RcNVUnavailable:  "the command may require writing of NV and NV is not current accessible",
+}
+
+// Indexes for arguments, handles and sessions.
+const (
+	Rc1 RcIndex = 0x01
+	Rc2         = 0x02
+	Rc3         = 0x03
+	Rc4         = 0x04
+	Rc5         = 0x05
+	Rc6         = 0x06
+	Rc7         = 0x07
+	Rc8         = 0x08
+	Rc9         = 0x09
+	RcA         = 0x0A
+	RcB         = 0x0B
+	RcC         = 0x0C
+	RcD         = 0x0D
+	RcE         = 0x0E
+	RcF         = 0x0F
+)
+
+type Error struct {
+	Code RcFmt0
+}
+
+func (e Error) Error() string {
+	return fmt.Sprintf("error code 0x%x : %s", e.Code, fmt0Msg[e.Code])
+}
+
+type VendorError struct {
+	Code uint32
+}
+
+func (e VendorError) Error() string {
+	return fmt.Sprintf("vendor error code 0x%x", e.Code)
+}
+
+type Warning struct {
+	Code RcWarn
+}
+
+func (w Warning) Error() string {
+	return fmt.Sprintf("warning code 0x%x : %s", w.Code, warnMgs[w.Code])
+}
+
+type ParameterError struct {
+	Code      RcFmt1
+	Parameter RcIndex
+}
+
+func (e ParameterError) Error() string {
+	return fmt.Sprintf("parameter %d, error code 0x%x : %s", e.Parameter, e.Code, fmt1Msg[e.Code])
+}
+
+type HandleError struct {
+	Code   RcFmt1
+	Handle RcIndex
+}
+
+func (e HandleError) Error() string {
+	return fmt.Sprintf("handle %d, error code 0x%x : %s", e.Handle, e.Code, fmt1Msg[e.Code])
+}
+
+type SessionError struct {
+	Code    RcFmt1
+	Session RcIndex
+}
+
+func (e SessionError) Error() string {
+	return fmt.Sprintf("session %d, error code 0x%x : %s", e.Session, e.Code, fmt1Msg[e.Code])
+}
+
+// Decode a TPM2 response code and return the appropriate error. Logic
+// according to the "Response Code Evaluation" chart in Part 1 of the TPM 2.0
+// spec.
+func decodeResponse(code tpmutil.ResponseCode) error {
+	if code == tpmutil.RCSuccess {
+		return nil
+	}
+	if code&0x180 == 0 { // Bits 7:8 == 0 is a TPM1 error
+		return fmt.Errorf("response status 0x%x", code)
+	}
+	if code&0x80 == 0 { // Bit 7 unset
+		if code&0x400 > 0 { // Bit 10 set, vendor specific code
+			return VendorError{uint32(code)}
+		}
+		if code&0x800 > 0 { // Bit 11 set, warning with code in bit 0:6
+			return Warning{RcWarn(code & 0x7f)}
+		}
+		// error with code in bit 0:6
+		return Error{RcFmt0(code & 0x7f)}
+	}
+	if code&0x40 > 0 { // Bit 6 set, parameter number in 8:11, code in 0:5
+		return ParameterError{RcFmt1(code & 0x3f), RcIndex((code & 0xf00) >> 8)}
+	}
+	if code&0x800 == 0 { // Bit 11 unset, handle in 8:10, code in 0:5
```

Comment:
"handle in 8:11" ?

Reply:
Is the question about the number of bits for the handle? The parameter# really is 4 bits, but handle# and session# in the other error types are only 3 bits as per TPM spec, part 1, figure 27